Consensus timer

jmmcgee opened the pull request:
Changes:

- Added `HeartbeatTimer` in order to periodically print simulation time passed to give user feedback during lock simulations
  - `HeartbeatTimer::start(..)` must be called after `sim.run(1)` 
- Added `Selector` to allow for probabilistic selection of peer to submit transactions to
  - In `ScaleFreeSim` transactions are submitted according to peers "ranks"

jmmcgee committed:
[FOLD] Added HeartbeatTimer to inform user of status:

HeartbeatTimer used in ScaleFreeSim_test

jmmcgee committed:
[FOLD] Add random weighted selector:

- use Selector in Submitter to send transactions probabilistcally to highest rank nodes

bachase commented on src/test/csf/Peer.h:
```diff
@@ -766,7 +766,7 @@ struct Peer
         // only reschedule if not completed
         if (completedLedgers < targetLedgers)
-            scheduler.in(parms().ledgerGRANULARITY, [&]() { timerEntry(); });
+            scheduler.in(parms().ledgerGRANULARITY, [this]() { this->timerEntry(); });
```
Style nit:  Rippled style does not use `this->` unless required, so I would just write ` [this]() { timerEntry(); }`

bachase commented on src/test/csf/SimTime.h:
```diff
@@ -27,6 +27,10 @@ namespace ripple {
 namespace test {
 namespace csf {
 
+using RealClock = std::chrono::system_clock;
```
I think these are useful, but perhaps the filename should change to Times.h?

Alternatively, until there is a need for `Real*` in more than one spot, perhaps it is better to leave them local to `timers.h`?

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+
+/** Gives heartbeat of simulation to signal simulation progression
+ */
+using namespace std::chrono;
```
`using namespace` in a header outside class or function scope is generally bad form, because anyone including the header ends up exposed to the namespaces definitions.

In this case, I would just move `using namespace std::chrono` inside your `beat` function.  For your default interval setting of `60s` you would just use the verbose `std::chrono::seconds{60}`.

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+    inline void
+    start()
+    {
+        scheduler_.in(interval_, [this](){this->beat(scheduler_.now());});
```
Same request to replace `this->beat(` with just `beat(`.

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+    {
+    };
+
+    inline void
```
Member functions defined in the class are implicitly `inline`.  Rippled style is to remove this `inline`.

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+        scheduler_.in(interval_, [this](){this->beat(scheduler_.now());});
+    };
+
+    inline void
```
As above, rippled style is to remove this `inline`.

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+                  << duration_cast<seconds>(realDuration).count()
+                  << "s}\n" << std::flush;
+
+        scheduler_.in(interval_, [this](){this->beat(scheduler_.now());});
```
Same request as above to drop `this->`

bachase commented on src/test/csf/timers.h:
```diff
@@ -0,0 +1,78 @@
+    inline void
+    start()
+    {
+        scheduler_.in(interval_, [this](){this->beat(scheduler_.now());});
```
Do you think it would be more useful to have the heartbeat based on real time elapsed rather than simulation time elapsed?  That is, every 10 seconds of realtime, report simulated time update?

jmmcgee replied:
I think it might be more useful but I wasn't sure how to do that. We would probably have to schedule another thread and have it sleep for the given time interval no? Didn't want to deal with that.

Also, after thinking about the general form of HeartbeatTimer and after talking to you, I think there is probably room for a generalization of simulation-time-processes such as HeartbeatTimer and Submitter. I imagine it would be a useful category to have in relationship to the simulation.

bachase replied:
Correct, it would likely require another thread.  There is already https://github.com/ripple/rippled/blob/develop/src/ripple/core/DeadlineTimer.h and an example in https://github.com/ripple/rippled/blob/develop/src/test/core/DeadlineTimer_test.cpp.

Just food for thought; your approach above is useful as is.

bachase commented on src/test/csf/random.h:
```diff
@@ -69,6 +69,39 @@ sample( std::size_t size, PDF pdf, Generator& g)
+template <class T, class Generator>
+class Selector
+{
+    std::vector<T>& v_;
```
Since this is meant to be generic code, I recommend generalizing away from `vector`.  Instead you could take begin/end iterators.  [Iterator traits](http://en.cppreference.com/w/cpp/iterator/iterator_traits) would help  and you could even static assert that the iterator has [`forward_iterator_tag`](http://en.cppreference.com/w/cpp/iterator/iterator_tags)

bachase commented on src/test/csf/random.h:
```diff
@@ -69,6 +69,39 @@ sample( std::size_t size, PDF pdf, Generator& g)
+    }
+
+    T&
+    operator()()
```
Do you need both `select()` and `operator()`? The general rule is to prefer one way to do things unless there is a strong need otherwise.

jmmcgee replied:
We do not need both. I was more confident in `select()`. Whenever I use operator overloading I'm a bit wary that maybe I haven't properly worked out the model so I have a tendency to implement the actual functionality separately and to simply pass through to it from the overload function. I can change it.

bachase commented on src/test/csf/submitters.h:
```diff
@@ -58,13 +58,13 @@ struct Rate
     SimTime stop_;
     std::uint32_t nextID_ = 0;
-    Peer & target_;
+    Selector & selector_;
```
What is the motivation for taking selector by reference versus by value?

jmmcgee replied:
No explicit motivation. Just didn't want to deal with working out the possible semantic issues as if `Selector` accidentally became not trivially copyable as I cludged things to work. In particular I was caught up in unrelated changes to `Selector` to allow it to be used with arbitrary `iterator`s to `T` instead of only taking a vector. As I unraveled those changes because I couldn't get them to work properly I simply forgot about using `Selector &`.

bachase reviewed (changes requested):
Looks really good.  Left some nits/comments that are mostly style related.  Not much I would change on the core design.

The travis build is failing because the visual studio project files are not updated.  Until we officially switch to cmake, the protocol is to run `scons vcxproj` and commit the changed files.  You can add that as an additional commit.

jmmcgee committed:
[FOLD] Fixed style conventions:

- do not use `this->` unless necessary
- do not pollute `namespace` with `using namespace` outside of function
  or class definition
- let `inline` modifier be implicit when function is defined in class
  definiton
- removed redundant `Selector::select()` in favor of `opeator()`

jmmcgee committed:
[FOLD] Selector takes random access iterators

jmmcgee committed:
[FOLD] Removed vector-specific

ghost commented:
yikes. accidentally deleted the branch and closed the PR. don't know fit their is a way to undo this. I re-`push`ed the branch of course. Sorry about that.

bachase commented:
Iterator changes look good.  I'll rebase later today after the CI builds come through.

I  think you could use forward iterators still and then its on the user to be aware of performance implications, but not worth caring about now.

jmmcgee committed:
[FOLD] updated visual studio build to pass CI

bachase commented:
Squashed and merged as f4d5035d4f6409e97cd855fc503f0479d868d0bb
